addons/mail/controllers/attachment.py

```diff
@@ -1,15 +1,49 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
+import io
+import logging
+import zipfile
+
 from werkzeug.exceptions import NotFound
 
 from odoo import _, http
 from odoo.exceptions import AccessError
-from odoo.http import request
+from odoo.http import request, content_disposition
+
 from odoo.tools import consteq
 from ..models.discuss.mail_guest import add_guest_to_context
 
+logger = logging.getLogger(__name__)
 
 class AttachmentController(http.Controller):
+
+    def _make_zip(self, name, attachments):
+        streams = (request.env['ir.binary']._get_stream_from(record, 'raw') for record in attachments)
+        # TODO: zip on-the-fly while streaming instead of loading the
+        #       entire zip in memory and sending it all at once.
```

Comment:
Added this comment because the zip on-the-fly could be improved to stream.
There are some libs to do this, like `python-zipstream`: https://github.com/allanlei/python-zipstream
In practice most files on messages are small, a few MBs, so I think this is decent.

Reply:
Point noted. Thanks a lot for the changes...